Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+  }
+  if (cfg.useFracPhiSlices < 0. || cfg.useFracPhiSlices >= 1.) {
+    ACTS_ERROR("value of useFracPhiSlices is "
+               << cfg.useFracPhiSlices
+               << ", allowed values are between 0 and 1");
+  }
+  if (std::isnan(cfg.numZSlices)) {
+    ACTS_ERROR("value of numZSlices was not initialized");
+  }
+  if (std::isnan(cfg.useFracZSlices)) {
+    ACTS_ERROR("value of useFracZSlices was not initialized");
+  }
+  if (cfg.useFracZSlices < 0. || cfg.useFracZSlices >= 1.) {
+    ACTS_ERROR("value of useFracZSlices is "
+               << cfg.useFracZSlices << ", allowed values are between 0 and 1");
+  }
+  if (std::isnan(cfg.maxAbsZ)) {
+    ACTS_ERROR("value of maxAbsZ was not initialized");
+  }
+  if (std::isnan(cfg.maxZPosition)) {
+    ACTS_ERROR("value of maxZPosition was not initialized");
+  }
+  if (std::isnan(cfg.maxRPosition)) {
+    ACTS_ERROR("value of maxRPosition was not initialized");
+  }
+  if (cfg.minimalizeWRT != "planes" && cfg.minimalizeWRT != "rays") {
+    ACTS_ERROR("value of minimalizeWRT is "
+               << cfg.minimalizeWRT
+               << ", allowed values are \"planes\" or \"rays\" ");
+  }
+  if (std::isnan(cfg.maxIterations)) {
+    ACTS_ERROR("value of maxIterations was not initialized");
+  }
+  if (std::isnan(cfg.removeFraction)) {
+    ACTS_ERROR("value of removeFraction was not initialized");
+  }
+  if (cfg.removeFraction < 0. || cfg.removeFraction >= 1.) {
+    ACTS_ERROR("value of removeFraction is "
+               << cfg.removeFraction << ", allowed values are between 0 and 1");
+  }
+  if (std::isnan(cfg.minVtxShift)) {
+    ACTS_ERROR("value of minVtxShift was not initialized");
+  }
+}
+
+template <typename spacepoint_t>
+Acts::Vector3 Acts::SeedVertexFinder<spacepoint_t>::findVertex(
+    const std::vector<spacepoint_t>& spacepoints) {
+  // sort spacepoints to different phi and z bins
+  std::vector<std::vector<std::vector<
+      std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>>>>>
+      sorted_spacepoints = sortSpacepoints(spacepoints);
+
+  // find triplets
+  std::vector<Acts::SeedVertexFinder<spacepoint_t>::Triplet> triplets =
+      findTriplets(sorted_spacepoints);
+
+  // if no valid triplets found
+  if (triplets.empty()) {
+    return {};
+  }
+
+  Acts::Vector3 vtx = Acts::Vector3::Zero();
+  if (m_cfg.minimalizeWRT == "planes") {
+    // find a point closest to all planes defined by the triplets
+    vtx = findClosestPointFromPlanes(triplets);
+  } else if (m_cfg.minimalizeWRT == "rays") {
+    // find a point closest to all rays fitted through the triplets
+    vtx = findClosestPointFromRays(triplets);
+  } else {
+    ACTS_ERROR("value of minimalizeWRT is "
+               << m_cfg.minimalizeWRT
+               << ", allowed values are \"planes\" or \"rays\" ");
+  }
+
+  return vtx;
+}
+
+template <typename spacepoint_t>
+std::vector<std::vector<
+    std::vector<std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>>>>>
```

Comment:
did not see east const in our codebase yet but also not sure if there is a guideline about it. anyways as long as it is consistent here I am fine with it

Reply:
I didn't want to copy the whole spacepoints but at the same time, I needed to put them into several vectors. Pointer to the constant worked well. There's no way the spacepoints themselves can be modified (and they shouldn't).
Also, it allows to modify the vectors later. Sorting the spacepoints according their phi might speed up the creation of the triplets later. I decided not include it in the first version as it will require more testing, however we are prepared for that.